test/e2e/hard_stop_after_test.go

```diff
@@ -0,0 +1,368 @@
+		t.Fatalf("some component still has hard-stop-after set: %v", err)
+	}
+}
+
+func hardStopAfterTestIngressConfig(t *testing.T, ingressConfig *configv1.Ingress, routerDeployment *appsv1.Deployment) error {
+	if err := setHardStopAfterDurationForIngressConfig(t, kclient, hardStopAfterRetryTimeout, hardStopAfterDurationPtr(hardStopAfterIngressConfigTimeout), ingressConfig); err != nil {
+		return fmt.Errorf("failed to update ingress config: %v", err)
+	}
+	if err := waitForIngressControllerCondition(t, kclient, 5*time.Minute, defaultName, defaultAvailableConditions...); err != nil {
+		return fmt.Errorf("failed to observe expected conditions: %v", err)
+	}
+	if err := waitForIngressConfigHardStopAfterToBeSet(kclient, hardStopAfterRetryTimeout, ingressConfig, hardStopAfterIngressConfigTimeout); err != nil {
+		return fmt.Errorf("hard-stop-after not set on ingress config: %v", err)
+	}
+	if err := waitForRouterDeploymentHardStopAfterToBeSet(kclient, routerDeployTimeout, routerDeployment, hardStopAfterIngressConfigTimeout); err != nil {
+		return fmt.Errorf("expected router deployment to have hard-stop-after configured: %v", err)
+	}
+	return nil
+}
+
+func hardStopAfterTestIngressController(t *testing.T, ic *operatorv1.IngressController, routerDeployment *appsv1.Deployment) error {
+	if err := setHardStopAfterDurationForIngressController(t, kclient, hardStopAfterRetryTimeout, hardStopAfterDurationPtr(hardStopAfterIngressControllerTimeout), ic); err != nil {
+		t.Fatalf("failed to update ingresscontroller: %v", err)
+	}
+	if err := waitForIngressControllerCondition(t, kclient, 5*time.Minute, defaultName, defaultAvailableConditions...); err != nil {
+		t.Fatalf("failed to observe expected conditions: %v", err)
+	}
+	if err := waitForIngressControllerHardStopAfterToBeSet(kclient, hardStopAfterRetryTimeout, ic, hardStopAfterIngressControllerTimeout); err != nil {
+		t.Fatalf("hard-stop-after not set on ingress controller: %v", err)
+	}
+	if err := waitForRouterDeploymentHardStopAfterToBeSet(kclient, routerDeployTimeout, routerDeployment, hardStopAfterIngressControllerTimeout); err != nil {
+		t.Fatalf("expected router deployment to have hard-stop-after configured: %v", err)
+	}
+	return nil
+}
+
+func WaitForHardStopAfterUnsetInAllComponents(client client.Client, timeout time.Duration, ic *operatorv1.IngressController, ingressConfig *configv1.Ingress, routerDeployment *appsv1.Deployment) error {
+	if err := waitForIngressConfigHardStopAfterToBeUnset(client, timeout, ingressConfig); err != nil {
+		return err
+	}
+	if err := waitForIngressControllerHardStopAfterToBeUnset(client, timeout, ic); err != nil {
+		return err
+	}
+	return waitForRouterDeploymentHardStopAfterToBeUnset(client, timeout, routerDeployment)
+}
+
+func waitForIngressConfigHardStopAfterToBeSet(client client.Client, timeout time.Duration, c *configv1.Ingress, expectedDuration time.Duration) error {
```

Comment:
could `waitForIngressConfigHardStopAfterToBeSet`, `waitForIngressConfigHardStopAfterToBeUnSet`, `waitForIngressControllerHardStopAfterToBeSet`, & `waitForIngressControllerHardStopAfterToBeUnSet` all be collapsed into one function with additional parameters for object (ingressconfig or ingresscontroller) and a set/unset bool?

Reply:
Funnily enough I did that first time; second time I found it easier to read with the discrete functions.